net/dns/wsl_windows.go

```diff
@@ -197,21 +198,75 @@ func (fs wslFS) WriteFile(name string, contents []byte, perm os.FileMode) error
 	cmd := fs.cmd("tee", "--", name)
 	cmd.Stdin = bytes.NewReader(contents)
 	cmd.Stdout = nil
-	if err := cmd.Run(); err != nil {
+	if err := wslRun(cmd); err != nil {
 		return err
 	}
-	return fs.cmd("chmod", "--", fmt.Sprintf("%04o", perm), name).Run()
+	return wslRun(fs.cmd("chmod", "--", fmt.Sprintf("%04o", perm), name))
 }
 
 func (fs wslFS) cmd(args ...string) *exec.Cmd {
 	cmd := wslCommand("-u", fs.user, "-d", fs.distro, "-e")
 	cmd.Args = append(cmd.Args, args...)
-	fmt.Printf("wslFS.cmd: %v\n", cmd.Args)
 	return cmd
 }
 
 func wslCommand(args ...string) *exec.Cmd {
 	cmd := exec.Command("wsl.exe", args...)
-	cmd.SysProcAttr = &syscall.SysProcAttr{HideWindow: true}
 	return cmd
 }
+
+func wslCombinedOutput(cmd *exec.Cmd) ([]byte, error) {
+	buf := new(bytes.Buffer)
+	cmd.Stdout = buf
+	cmd.Stderr = buf
+	err := wslRun(cmd)
+	return buf.Bytes(), err
+}
+
+func wslRun(cmd *exec.Cmd) (err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("wslRun(%v): %v", cmd.Args, err)
+		}
+	}()
+
+	var token syscall.Token
+	if u, err := user.Current(); err == nil && u.Name == "SYSTEM" {
+		// We need to switch user to run wsl.exe.
+		// https://github.com/microsoft/WSL/issues/4803
+		sessionID, err := wtsGetActiveConsoleSessionID()
+		if err != nil {
+			return err
+		}
+
+		var t windows.Token
+		if err := windows.WTSQueryUserToken(sessionID, &t); err != nil {
+			return err
+		}
+		defer t.Close()
+		token = syscall.Token(token)
+	}
+
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		Token:      token,
+		HideWindow: true,
+	}
+	return cmd.Run()
+}
+
+var (
+	kernel32                         = syscall.NewLazyDLL("kernel32.dll")
+	procWTSGetActiveConsoleSessionId = kernel32.NewProc("WTSGetActiveConsoleSessionId")
+)
+
+// wtsGetActiveConsoleSessionID Returns the session ID of the console session.
+//
+// https://msdn.microsoft.com/en-us/library/aa383835(VS.85).aspx
+// DWORD WTSGetActiveConsoleSessionId();
+func wtsGetActiveConsoleSessionID() (sessionID uint32, err error) {
```

Comment:
move this to util/winutil? could even do the //sys thing to auto-generate the Go wrapper. (e.g. see https://github.com/tailscale/corp/commit/8a4f261c489d842049f8d8a9cdd3d1571cb0c428 for using the syscall generator)

Or really, we can add it to x/sys/windows tonight too.

Reply:
I tried putting this in x/sys/windows using the `//sys` wrapper but it generates the wrong code:

```
func WTSGetActiveConsoleSessionId() (sessionID uint32, err error) {
       r0, _, e1 := syscall.Syscall(procWTSGetActiveConsoleSessionId.Addr(), 0, 0, 0, 0)
       sessionID = uint32(r0)
       if sessionID == 0 {
               err = errnoErr(e1)
       }
       return
}
```

The documentation linked above says 'If there is no session attached to the physical console, (for example, if the physical console session is in the process of being attached or detached), this function returns 0xFFFFFFFF.'

So I guess I'll move it to winutil.

Reply:
I don't know if anyone cares about x/sys having the correct semantics baked in or not, so I spent an hour figuring out how to get gerrit out of mothballs and sent https://go-review.googlesource.com/c/sys/+/331909, figuring I can do the 0xffffff check myself.